### PR TITLE
Unified docker image for BIM edition. Add hooks for Cloud platform

### PR DESCRIPTION
**Pull request by crohr**
Simplifies Dockerfile by adding BIM dependencies as default, and adds hooks for the various platforms (on-premise, cloud).

Goal being to unify our various Dockerfiles to simplify CI/CD.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -6,83 +6,57 @@ RUN /tmp/build-pgloader && rm /tmp/build-pgloader
...
 
-ENV NODE_VERSION "10.15.0"
-ENV BUNDLER_VERSION "2.0.2"
+# Allow platform-specific additions (e.g. cloud)
+ARG PLATFORM=on-premise
+# Use OAuth token from read-only openprojectdeploy user to fetch private gems
+ARG GITHUB_OAUTH_TOKEN
+ARG DEBIAN_FRONTEND=noninteractive
+
+ENV NODE_VERSION="10.15.0"
```
Comment (oliverguenther) [inline Dockerfile]:
```suggestion
ENV NODE_VERSION="12.18.3"
```

File: docker/setup/postinstall.sh [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+echo " ---> POSTINSTALL"
+
+# Add MySQL-to-Postgres migration script to path (used in entrypoint.sh)
+cp ./docker/mysql-to-postgres/bin/migrate-mysql-to-postgres /usr/local/bin/
+
+# Ensure we can write in /tmp/op_uploaded_files (cf. #29112)
+mkdir -p /tmp/op_uploaded_files/ && chown -R $APP_USER:$APP_USER /tmp/op_uploaded_files/
+
+cp ./packaging/conf/database.yml ./config/database.yml
```
Comment (crohr) [inline docker/setup/postinstall.sh]: @oliverguenther is there any reason to still have the current `config/database.yml` in the repo? I understand mysql support is deprecated anyway?

File: Dockerfile [hunk 3]
```diff
@@ -6,83 +6,58 @@ RUN /tmp/build-pgloader && rm /tmp/build-pgloader
...
+ENV APP_DATA_PATH_LEGACY=/var/db/openproject
+ENV PGBIN="/usr/lib/postgresql/9.6/bin"
+ENV PGDATA=/var/openproject/pgdata
+ENV PGDATA_LEGACY=/var/lib/postgresql/9.6/main
+
+ENV DATABASE_URL=postgres://openproject:openproject@127.0.0.1/openproject
+ENV HEROKU=true
+ENV RAILS_ENV=production
+ENV RAILS_CACHE_STORE=memcache
+ENV RAILS_GROUPS="production docker opf_plugins"
```
Comment (machisuji) [inline Dockerfile:30]: `BUNDLER_GROUPS` would be a more fitting name here as this is unrelated to rails.

File: docker/setup/postinstall.sh (hunk 2 above)
Comment (machisuji) [reply docker/setup/postinstall.sh]: The encoding option may still be relevant I reckon. But we surely can remove the MySQL bit from the packager `database.yml`.

File: Dockerfile [hunk 4]
```diff
@@ -6,83 +6,58 @@ RUN /tmp/build-pgloader && rm /tmp/build-pgloader
 FROM ruby:2.6-stretch
 MAINTAINER operations@openproject.com
 
-ENV NODE_VERSION "10.15.0"
-ENV BUNDLER_VERSION "2.0.2"
+# Allow platform-specific additions. Valid values are: on-premise,cloud
+ARG PLATFORM=on-premise
```
Comment (machisuji) [inline Dockerfile]: I know it's nitpicky but as I just learned not too long ago the proper term is apparently `on-premises` rather than `on-premise`. It doesn't really matter seeing as we're just in a build file here but we might as well do it right.

**Review by machisuji — COMMENTED** [review]
Nice work! Looks good. I'm still testing this. Right now the seed is failing for the built docker image (works outside of docker with the same code). I'm still looking into that and see what the issue is before we can merge.

**Comment by machisuji** [comment]
Can I suggest an improvement to the docker image while we're a it? For debugging purposes it would be extremely helpful if one was able to get a shell as `root` on the container rather than `app` when running `docker run -it openproject/community:latest bash`.

Do you have any objections to changing the user from `app` to `root` bash too?

File: Dockerfile [hunk 5]
```diff
@@ -6,83 +6,58 @@ RUN /tmp/build-pgloader && rm /tmp/build-pgloader
...
+# Use OAuth token in case private gems need to be fetched
+ARG GITHUB_OAUTH_TOKEN
+ARG DEBIAN_FRONTEND=noninteractive
+
+ENV NODE_VERSION="12.18.3"
+ENV BUNDLER_VERSION="2.0.2"
```
Comment (machisuji) [inline Dockerfile:16]: I suggest we use `2.1.4` here to keep up with what's defined in the `Gemfile.lock` already.

File: Dockerfile [hunk 6]
```diff
@@ -6,83 +6,58 @@ RUN /tmp/build-pgloader && rm /tmp/build-pgloader
...
+ENV PGDATA_LEGACY=/var/lib/postgresql/9.6/main
+
+ENV DATABASE_URL=postgres://openproject:openproject@127.0.0.1/openproject
+ENV HEROKU=true
+ENV RAILS_ENV=production
+ENV RAILS_CACHE_STORE=memcache
+ENV RAILS_GROUPS="production docker opf_plugins"
+ENV OPENPROJECT_INSTALLATION__TYPE=docker
+# Valid values are: default,bim
+ENV OPENPROJECT_EDITION=default
```
Comment (machisuji) [inline Dockerfile]: This has to be `standard`, not `default`. Otherwise the seeding will fail. I'll go ahead and push the fix.

**Review by machisuji — APPROVED** [review]
Good to merge now.

**Comment by machisuji** [comment]
Thanks @crohr , this is great. If you agree with my allowing root access for the bash command which I pushed (alongside 1 bugfix and the renaming of the RAILS_GROUPS to BUNDLER_GROUPS) this can be merged.

File: Dockerfile (hunk 3 above)
Comment (crohr) [reply Dockerfile:30]: Actually, `RAILS_GROUPS` is the env variable supported by Rails to load additional groups (https://api.rubyonrails.org/classes/Rails.html#method-c-groups).

**Comment by crohr** [comment]
> Can I suggest an improvement to the docker image while we're a it? For debugging purposes it would be extremely helpful if one was able to get a shell as `root` on the container rather than `app` when running `docker run -it openproject/community:latest bash`.
> 
> Do you have any objections to changing the user from `app` to `root` bash too?

@machisuji you can pass `-u root` when starting a container with `docker run` so I'm not sure it makes sense to remove a bit of security for that. Also when you exec into a container, you end up as the root user.

Edit: actually I understand what you're saying, with the entrypoint stuff. Not sure about hardcoding `bash` as executing under the `root` user since you may want to investigate what's in the container as the `app` user. Maybe add a filter on a specific command (`root-bash` or something) that would trigger a `bash` as root?

File: docker/setup/postinstall.sh [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+	export NG_CLI_ANALYTICS=ci # so angular cli doesn't block waiting for user input
+
+	# Installing frontend dependencies
+	RAILS_ENV=production npm install
+
+	popd
+
+	# Bundle assets
+	su - postgres -c "$PGBIN/initdb -D /tmp/nulldb"
```
Comment (crohr) [inline docker/setup/postinstall.sh:39]: FYI this change means we can get rid of the dynamically provisioned DB we use in some of our docker compilation jobs.

File: Dockerfile (hunk 4 above)
Comment (crohr) [reply Dockerfile]: Maybe use `on-prem` then, which seems widely used as well (maybe for that reason)?

File: Dockerfile [hunk 8]
```diff
@@ -6,83 +6,58 @@ RUN /tmp/build-pgloader && rm /tmp/build-pgloader
...
+ENV OPENPROJECT_INSTALLATION__TYPE=docker
+# Valid values are: standard,bim
+ENV OPENPROJECT_EDITION=standard
+ENV NEW_RELIC_AGENT_ENABLED=false
+ENV ATTACHMENTS_STORAGE_PATH=$APP_DATA_PATH/files
 # Set a default key base, ensure to provide a secure value in production environments!
-ENV SECRET_KEY_BASE OVERWRITE_ME
+ENV SECRET_KEY_BASE=OVERWRITE_ME
 
 COPY --from=pgloader /usr/local/bin/pgloader-ccl /usr/local/bin/
```
Comment (crohr) [inline Dockerfile:39]: @oliverguenther @machisuji do you have an idea when we might remove pgloader from the docker image? How long has it been since the switch from MySQL?

Comment (oliverguenther) [reply Dockerfile:39]: As discussed, I would vote to leave it in until we remove the respective migration tools. If it's taking too much space, that is

File: docker/setup/postinstall.sh (hunk 2 above)
Comment (crohr) [reply docker/setup/postinstall.sh]: We'll just remove `config/database.yml` so that everything comes from the `DATABASE_URL` env variable.

**Comment by crohr** [comment]
Updated the Docker workflow to also test the image after pushing it. Couldn't test it locally with `act` so we'll see if that works once it's merged.
